CMakeLists.txt

```diff
@@ -0,0 +1,403 @@
+if(WITH_EXTENUM_FUNC)
+    set(EXTENUM_FUNC ${WITH_EXTENUM_FUNC})
+endif()
+
+if(EXTENUM_LIB)
+    if(NOT EXTENUM_FUNC)
+        set(EXTENUM_FUNC ${EXTENUM_LIB})
+    endif()
+    try_compile(EXTENUM_LINK_SUCCESS ${CMAKE_BINARY_DIR} SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/extenum_check.cpp
+        LINK_LIBRARIES ${EXTENUM_LIB}
+    )
+    if(NOT EXTENUM_LINK_SUCCESS)
+        message(FATAL_ERROR "Failed to find external enumeration")
+    endif()
+    set(EXTENUM_LIBS "-l${EXTENUM_LIB}")
+    add_definitions(-DFPLLL_EXTENUM_FUNC=${EXTENUM_FUNC})
+endif()
+set(EXTENUM_LIBS "${EXTENUM_LIBS}" CACHE STRING "External enumeration libraries")
+
+# Include directories
+include_directories(${GMP_INCLUDE_DIR} ${MPFR_INCLUDE_DIR} ${QD_INCLUDE_DIR})
+set(LIBS "${LIBGMP_LIB} ${LIBMPFR_LIB} ${LIBQD_LIB} ${PTHREAD_LIB}")
+message(STATUS "LIBS: ${LIBS}")
+
+# setup fplll/fplll_config.h
+configure_file(
+  ${CMAKE_CURRENT_SOURCE_DIR}/fplll/fplll_config.cmake.h.in
+  ${CMAKE_CURRENT_SOURCE_DIR}/fplll/fplll_config.h
+  @ONLY)
+
+# setup ./config.h
+include(CheckIncludeFileCXX)
+include(CheckFunctionExists)
+include(CheckTypeSize)
+set(CMAKE_REQUIRED_FLAGS "-isysroot ${CMAKE_SYSROOT}")
+check_include_file_cxx(cstdio HAVE_CSTDIO)
+check_include_file_cxx(iostream HAVE_IOSTREAM)
+check_include_file_cxx(string HAVE_STRING)
+check_include_file_cxx(vector HAVE_VECTOR)
+check_include_file_cxx(cstdint HAVE_CXX11)
+check_include_file_cxx(limits HAVE_LIMITS)
+check_include_file_cxx(limits.h HAVE_LIMITS_H)
+check_include_file_cxx(dlfcn.h HAVE_DLFCN_H)
+check_include_file_cxx(inttypes.h HAVE_INTTYPES_H)
+check_include_file_cxx(stdbool.h HAVE_STDBOOL_H)
+check_include_file_cxx(stdint.h HAVE_STDINT_H)
+check_include_file_cxx(stdio.h HAVE_STDIO_H)
+check_include_file_cxx(stdlib.h HAVE_STDLIB_H)
+check_include_file_cxx(strings.h HAVE_STRINGS_H)
+check_include_file_cxx(string.h HAVE_STRING_H)
+check_include_file_cxx(sys/stat.h HAVE_SYS_STAT_H)
+check_include_file_cxx(sys/types.h HAVE_SYS_TYPES_H)
+check_include_file_cxx(unistd.h HAVE_UNISTD_H)
+check_function_exists(floor HAVE_FLOOR)
+check_function_exists(pow HAVE_POW)
+check_function_exists(rint HAVE_RINT)
+check_function_exists(sqrt HAVE_SQRT)
+check_function_exists(strtol HAVE_STRTOL)
+check_type_size("_Bool" HAVE__BOOL LANGUAGE C)
+configure_file(
+  ${CMAKE_CURRENT_SOURCE_DIR}/config.cmake.h.in
+  ${CMAKE_CURRENT_SOURCE_DIR}/config.h
+  @ONLY)
+
+# Set up fplll package, corresponding to fplll/Makefile.am and ./fplll.pc.in
+# NOTE: ignored distribution related stuff, EXTRA_DIST not translated over
+include(GNUInstallDirs)
+
+set(FPLLL_PKG_CONFIG "${CMAKE_CURRENT_SOURCE_DIR}/fplll.pc")
+# setup fplll.pc for package
+configure_file(
+  ${CMAKE_CURRENT_SOURCE_DIR}/fplll.cmake.pc.in
+  ${FPLLL_PKG_CONFIG}
+  @ONLY)
+
+# Compiler flags
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -O3")
+## enabled via -DDEBUG
+if(DEBUG)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g")
```

Comment:
I feel like this is confusing because of the clash with `-DNDEBUG`. That is, I could write:

```
cmake -DNDEBUG -DDEBUG ..
```

And it would produce an executable with assertions disabled but debug information in the executable. I think that's somewhat counter-intuitive: could we perhaps use a different name?

Reply:
very good point. maybe I'll directly `if (not NDEBUG)` instead?